### PR TITLE
Navigation between pages and books with history

### PR DESCRIPTION
**Pull request by hatton**
This change is [Reviewable](https://reviewable.io/reviews/BloomBooks/bloom-player/347)

## Summary by CodeRabbit

- **New Features**
  - Enhanced Storybook configuration for improved static file handling and story sorting.
  - Introduced new styles for multilingual book layouts and branding elements.
  - Added new JSON configuration files for appearance settings and user preferences.
  - New `BloomPlayerIframe` component and `useBuildIsReady` hook for iframe integration.

- **Bug Fixes**
  - Improved back button functionality and visibility management in the Bloom Player.

- **Documentation**
  - Updated README files for clarity on usage and integration of the bloom-player package.
  
- **Tests**
  - Comprehensive unit tests added for navigation functions and interaction within the iframe.

- **Style**
  - New CSS files and classes added for better layout and presentation across various book formats.

### CODE REVIEW

**Comment by andrew-polk** [comment]
*[`src/bloom-player-core.tsx` line 1689 at r3](https://reviewable.io/reviews/BloomBooks/bloom-player/347#-ODl40cs9WGLxTrUKFL2:-ODl40cs9WGLxTrUKFL3:bjzmlqt) ([raw file](https://github.com/BloomBooks/bloom-player/blob/04a01f1ad8372c1a3063f83572f0981a713bd965/src/bloom-player-core.tsx#L1689)):*
> ```XML
>     // a corresponding woff font in a standard location. We may even be able to pay to
>     // provide some fonts there for book previewing that are NOT embeddable.
>     private makeFontStylesheet(href: string): void {
> ```

I think all this got readded by accident during a merge?
You had previously pulled it out to a separate file.

File: .storybook/main.ts [hunk 1]
```diff
@@ -18,15 +20,49 @@ const config: StorybookConfig = {
...
+            // so long as the html file of the book is named "index.htm".
+            "/book/": {
+                target: `http://localhost:6006`,
+                rewrite: (path) => {
+                    const bookId = path.split("/")[2];
+                    const bookName = idToBookName[bookId];
+                    const rewrittenPath = path.replace(
+                        `/book/${bookId}`,
+                        `testBooks/${bookName}`,
+                    );
```
[the remark this answers is not in the document]
Comment (hatton) [reply .storybook/main.ts:44]: Skip

File: .storybook/main.ts [hunk 2]
```diff
@@ -18,15 +20,49 @@ const config: StorybookConfig = {
...
+                    console.log(`proxy got request ${path}`);
+                    return path.replace("/bloom/", "");
+                },
+            },
+        };
+        return {
+            ...config,
+            server: { ...config.server, proxy: updatedProxy },
+            addons: ["@storybook/addon-interactions"],
+        };
```
[the remark this answers is not in the document]
Comment (hatton) [reply .storybook/main.ts:65]: Skip

File: public/testBooks/multibook-index/appearance.css [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+ .A4Landscape,
+ .A4Portrait,
+ .B5Portrait,
+ .LetterLandscape,
+ .LetterPortrait,
+ .LegalLandscape,
+ .LegalLandscape {
+    --page-margin: 15mm;
+    --page-gutter: 10mm;
+}
```
[the remark this answers is not in the document]
Comment (hatton) [reply public/testBooks/multibook-index/appearance.css:140]: Skip

File: public/testBooks/multibook-index/Basic Book.css [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+}
+.nospecidalmovement_imageOnTop .bloom-editable.bloom-monolingual {
+  height: 100%;
+}
+.bloom-bilingual.nospecidalmovement_imageOnTop .bloom-editable {
+  height: calc(50% - var(--multilingual-editable-vertical-gap));
+}
+.bloom-trilingual.nospecidalmovement_imageOnTop .bloom-editable {
+  height: calc(33% - var(--multilingual-editable-vertical-gap));
+}
```
[the remark this answers is not in the document]
Comment (hatton) [reply public/testBooks/multibook-index/Basic Book.css:87]: skip

File: public/testBooks/multibook-index/branding.css [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+.Cm13Landscape .landscapeOnly,
+.Device16x9Landscape .landscapeOnly,
+.LetterLandscape .landscapeOnly,
+.LegalLandscape .landscapeOnly,
+.A3Landscape .landscapeOnly,
+.A4Landscape .landscapeOnly,
+.A5Landscape .landscapeOnly,
+.A6Landscape .landscapeOnly {
+  display: unset;
+}
```
[the remark this answers is not in the document]
Comment (hatton) [reply public/testBooks/multibook-index/branding.css:54]: skip

File: public/testBooks/multibook-index/branding.css [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+.outsideBackCover img.branding {
+  margin-top: 4px;
+  width: 10cm;
+}
```
[the remark this answers is not in the document]
Comment (hatton) [reply public/testBooks/multibook-index/branding.css:58]: skip

File: public/testBooks/multibook-index/index.htm [hunk 7]
```diff
@@ -0,0 +1,473 @@
+﻿<!DOCTYPE html>
+
+<html>
+
+<head>
+    <meta charset="UTF-8">
+    </meta>
+    <meta name="Generator" content="Bloom Version 6.2.0 (apparent build date: 27-Nov-2024)">
```
[the remark this answers is not in the document]
Comment (hatton) [reply public/testBooks/multibook-index/index.htm:8]: skip

File: public/testBooks/multibook-index/index.htm [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+        <div class="marginBox">
+            <div data-book="cover-branding-top-html" lang="*"></div>
+
+            <div class="bloom-translationGroup bookTitle" data-default-languages="V,N1"
+                data-visibility-variable="cover-title-LN-show" style="font-size: 33.3333px;">
+                <label class="bubble">Book title in {lang}</label>
+
+                <div class="bloom-editable bloom-nodefaultstylerule Title-On-Cover-style bloom-padForOverflow bloom-visibility-code-on bloom-content1 bloom-contentNational1 bloom-contentFirst"
+                    lang="en" contenteditable="true" data-book="bookTitle" style="padding-bottom: 0px;" tabindex="0"
+                    spellcheck="false" role="textbox" aria-label="false" data-hasqtip="true" aria-describedby="qtip-0">
```
[the remark this answers is not in the document]
Comment (hatton) [reply public/testBooks/multibook-index/index.htm:150]: skip

File: public/testBooks/multibook-index/index.htm [hunk 9]
```diff
@@ -0,0 +1,473 @@
...
+                    <p>multibook-index</p>
+                </div>
+
+                <div class="bloom-editable bloom-nodefaultstylerule Title-On-Cover-style bloom-padForOverflow" lang="z"
+                    contenteditable="true" data-book="bookTitle"></div>
+            </div>
+
+            <div class="bloom-imageContainer">
+                <img data-book="coverImage" src="placeHolder.png" alt=""></img>
```
[the remark this answers is not in the document]
Comment (hatton) [reply public/testBooks/multibook-index/index.htm:159]: skip

File: public/testBooks/multibook-index/index.htm [hunk 10]
```diff
@@ -0,0 +1,473 @@
...
+                        <p>This is a page in the index book that has links to another book</p>
+                        <p><a href="#cover">This book Cover</a></p>
+
+                        <p><a href="#0b22a586-3324-4a76-b95d-e49ca03105d8">This book pirate page</a></p>
+
+                        <p><a href="/book/2c1b71ac-f399-446d-8398-e61a8efd4e83">Book 1</a>
+                        </p>
+                        <p><a href="/book/2c1b71ac-f399-446d-8398-e61a8efd4e83#b0f5224f-7a31-4944-84fb-f2cd46d56fc8">Book
+                                1 Crab Page</a>
+                        </p>
```
[the remark this answers is not in the document]
Comment (hatton) [reply public/testBooks/multibook-index/index.htm:371]: skip

File: public/testBooks/multibook-target1/appearance.css [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+ .A4Landscape,
+ .A4Portrait,
+ .B5Portrait,
+ .LetterLandscape,
+ .LetterPortrait,
+ .LegalLandscape,
+ .LegalLandscape {
+    --page-margin: 15mm;
+    --page-gutter: 10mm;
+}
```
[the remark this answers is not in the document]
Comment (hatton) [reply public/testBooks/multibook-target1/appearance.css:140]: skip

**Review by hatton — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/bloom-player/347)* status: 48 of 62 files reviewed, 32 unresolved discussions (waiting on @andrew-polk and @coderabbitai[bot])

___
*[`.storybook/main.ts` line 11 at r2](https://reviewable.io/reviews/BloomBooks/bloom-player/347#-ODfzoWj6scMJwXS_ZwG:-ODlN71q7qYUGN0QvXeM:bf64306) ([raw file](https://github.com/BloomBooks/bloom-player/blob/2226513713431b64df219adc8aa47f77922a00e1/.storybook/main.ts#L11)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Should there be a reference to this in the README?
</blockquote></details>

Personally I consider `package.json` to be the index to these things in itself. No value is added (but risk increased) by documenting in two places.
___
*[`public/testBooks/multibook-index/index.htm` line 1 at r2](https://reviewable.io/reviews/BloomBooks/bloom-player/347#-ODg-L0sBYOay_jkT4MH:-ODlOzMj-DN2x1_L3sUr:b88fe6h) ([raw file](https://github.com/BloomBooks/bloom-player/blob/2226513713431b64df219adc8aa47f77922a00e1/public/testBooks/multibook-index/index.htm#L1)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Note that I did not really review the files in `public/testBooks/multibook-index{1}` but there are certainly some we could remove such as 
- history.db 
- userprefs
- thumbnail
- distribution
- publish-settings
</blockquote></details>

done except `distribution`, which bloom-player reads

File: public/testBooks/multibook-target1/Basic Book.css [hunk 12]
```diff
@@ -0,0 +1,348 @@
...
+}
+.nospecidalmovement_imageOnTop .bloom-editable.bloom-monolingual {
+  height: 100%;
+}
+.bloom-bilingual.nospecidalmovement_imageOnTop .bloom-editable {
+  height: calc(50% - var(--multilingual-editable-vertical-gap));
+}
+.bloom-trilingual.nospecidalmovement_imageOnTop .bloom-editable {
+  height: calc(33% - var(--multilingual-editable-vertical-gap));
+}
```
[the remark this answers is not in the document]
Comment (hatton) [reply public/testBooks/multibook-target1/Basic Book.css:87]: skip

File: public/testBooks/multibook-target1/Basic Book.css [hunk 13]
```diff
@@ -0,0 +1,348 @@
...
+  padding-top: 10px;
+}
+.pictureAndWordPage .bloom-editable {
+  vertical-align: middle;
+  height: auto;
+}
+.pictureAndWordPage .bloom-content2,
+.pictureAndWordPage .bloom-content3 {
+  font-size: 200%;
+}
```
[the remark this answers is not in the document]
Comment (hatton) [reply public/testBooks/multibook-target1/Basic Book.css:347]: we don't want to review html or css in this pr

File: public/testBooks/multibook-target1/basePage.css [hunk 14]
```diff
@@ -0,0 +1,1304 @@
...
+}
+.bloom-vertical-align-center:not(:has(.overflow)):not(.bloom-bilingual):not(.bloom-trilingual) .bloom-editable {
+  justify-content: center;
+}
+.bloom-vertical-align-bottom:not(:has(.overflow)) {
+  justify-content: flex-end;
+}
+.bloom-vertical-align-bottom:not(:has(.overflow)):not(.bloom-bilingual):not(.bloom-trilingual) .bloom-editable {
+  justify-content: flex-end;
+}
```
[the remark this answers is not in the document]
Comment (hatton) [reply public/testBooks/multibook-target1/basePage.css:918]: we don't want to review html or css in this pr

File: public/testBooks/multibook-target1/branding.css [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+.A3Landscape .landscapeOnly,
+.A4Landscape .landscapeOnly,
+.A5Landscape .landscapeOnly,
+.A6Landscape .landscapeOnly {
+  display: unset;
+}
+.outsideBackCover img.branding {
+  margin-top: 4px;
+  width: 10cm;
+}
```
[the remark this answers is not in the document]
Comment (hatton) [reply public/testBooks/multibook-target1/branding.css:58]: we don't want to review html or css in this pr

File: public/testBooks/multibook-target1/index.htm [hunk 16]
```diff
@@ -0,0 +1,432 @@
...
+                </div>
+                <div class="split-pane-divider vertical-divider"></div>
+
+                <div class="split-pane-component position-right">
+                    <div class="split-pane-component-inner" min-height="60px 150px 250px" min-width="60px 150px"
+                        style="position: relative;">
+                        <div class="bloom-imageContainer bloom-leadingElement"
+                            data-title="Name: image.png Size: 35.15 kb Dots: 330 x 330 For the current paper size: • The image container is 324 x 328 dots. • For print publications, you want between 300-600 DPI (Dots Per Inch). ⚠ This image would print at 98 DPI. • An image with 1013 x 1025 dots would fill this container at 300 DPI."
+                            title=""><img src="image.png" alt="" data-copyright="" data-creator=""
+                                data-license=""></img></div>
```
[the remark this answers is not in the document]
Comment (hatton) [reply public/testBooks/multibook-target1/index.htm:374]: we don't want to review html or css in this pr

File: public/testBooks/multibook-target1/previewMode.css [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+  body {
+    background-color: #1a1a1a;
+  }
+}
+div.bloom-page.coverColor {
+  background-color: #d8adad;
+}
+html {
+  cursor: not-allowed;
+}
```
[the remark this answers is not in the document]
Comment (hatton) [reply public/testBooks/multibook-target1/previewMode.css:35]: we don't want to review html or css in this pr

File: public/testBooks/multibook-target1/Traditional-XMatter.css [hunk 18]
```diff
@@ -0,0 +1,262 @@
...
+.frontCover .bloom-translationGroup.bookTitle .bloom-editable.bloom-contentSecond {
+  order: 1;
+}
+.frontCover .bloom-translationGroup.bookTitle .bloom-editable.bloom-contentThird {
+  order: 2;
+}
+.frontCover .placeholder,
+.frontCover .bloom-imageContainer {
+  order: kFrontCoverImageOrder;
+  flex: 1 1 auto;
```
[the remark this answers is not in the document]
Comment (hatton) [reply public/testBooks/multibook-target1/Traditional-XMatter.css:53]: we don't want to review html or css in this pr

File: public/testBooks/multibook-target1/Traditional-XMatter.css [hunk 19]
```diff
@@ -0,0 +1,262 @@
...
+}
+.ISBNContainer .bloom-translationGroup {
+  display: inline-block;
+}
+.ISBNContainer .bloom-editable {
+  text-align: left !important;
+  width: 4.3cm !important;
+  max-height: 1.8em;
+  min-height: 1.8em;
+}
```
[the remark this answers is not in the document]
Comment (hatton) [reply public/testBooks/multibook-target1/Traditional-XMatter.css:255]: we don't want to review html or css in this pr

**Review by hatton — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/bloom-player/347)* status: 48 of 62 files reviewed, 32 unresolved discussions (waiting on @andrew-polk and @coderabbitai[bot])

File: .storybook/main.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (hatton) [reply .storybook/main.ts:44]: Done.

File: .storybook/main.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (hatton) [reply .storybook/main.ts:65]: Done.

File: public/testBooks/multibook-index/appearance.css (hunk 3 above)
[the remark this answers is not in the document]
Comment (hatton) [reply public/testBooks/multibook-index/appearance.css:140]: Done.

File: public/testBooks/multibook-index/Basic Book.css (hunk 4 above)
[the remark this answers is not in the document]
Comment (hatton) [reply public/testBooks/multibook-index/Basic Book.css:87]: Done.

File: public/testBooks/multibook-index/branding.css (hunk 5 above)
[the remark this answers is not in the document]
Comment (hatton) [reply public/testBooks/multibook-index/branding.css:54]: Done.

File: public/testBooks/multibook-index/branding.css (hunk 6 above)
[the remark this answers is not in the document]
Comment (hatton) [reply public/testBooks/multibook-index/branding.css:58]: Done.

File: public/testBooks/multibook-index/index.htm (hunk 7 above)
[the remark this answers is not in the document]
Comment (hatton) [reply public/testBooks/multibook-index/index.htm:8]: Done.

File: public/testBooks/multibook-index/index.htm (hunk 8 above)
[the remark this answers is not in the document]
Comment (hatton) [reply public/testBooks/multibook-index/index.htm:150]: Done.

File: public/testBooks/multibook-index/index.htm (hunk 9 above)
[the remark this answers is not in the document]
Comment (hatton) [reply public/testBooks/multibook-index/index.htm:159]: Done.

**Review by hatton — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/bloom-player/347)* status: 48 of 62 files reviewed, 32 unresolved discussions (waiting on @andrew-polk and @coderabbitai[bot])

**Review by andrew-polk — COMMENTED** [review]
> Dismissed @coderabbitai[bot] from 19 discussions.
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/bloom-player/347)* status: 45 of 62 files reviewed, 29 unresolved discussions (waiting on @coderabbitai[bot] and @hatton)

___
*[`.storybook/main.ts` line 11 at r2](https://reviewable.io/reviews/BloomBooks/bloom-player/347#-ODfzoWj6scMJwXS_ZwG:-ODw-Zz72QPF2I5YiW62:bckkpxu) ([raw file](https://github.com/BloomBooks/bloom-player/blob/2226513713431b64df219adc8aa47f77922a00e1/.storybook/main.ts#L11)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

Personally I consider `package.json` to be the index to these things in itself. No value is added (but risk increased) by documenting in two places.
</blockquote></details>

Makes sense to me. I brought it up because the README does already refer to running yarn storybook. And it seems like this might be necessary to really make that experience useful? But the dev will discover it eventually.

**Review by andrew-polk — COMMENTED** [review]
> Reviewed 2 of 62 files at r1, 4 of 14 files at r5.
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/bloom-player/347)* status: 51 of 62 files reviewed, 29 unresolved discussions (waiting on @coderabbitai[bot] and @hatton)

**Comment by andrew-polk** [comment]
Looks like this has been merged as https://github.com/BloomBooks/bloom-player/commit/fe80e7d03b79190e8b7aa258948a81158ef4549b

**Pull request closed without merging**